File: crates/testrunner/harness/java/check.java
// Vybe test harness — Java.
//
// Real Java, exactly as test262's `assert.js` is real JavaScript: this file
// compiles with `javac` on its own, so it can be read, formatted and debugged
// with Java's own tools.
//
// Java has no top-level functions, so the harness must be a static member of
// the test's class rather than prepended to the file. The emitter splices the
// members between the outermost braces below into `Main`, which is also where
// `run_main`/`run_in_main` put the program.
//
// Output is COLLECTED, not paired. The emitter rewrites every
// `System.out.println(x)` into `__p(x)` and compares the whole output once, so
// a program whose print count is not static — a loop — still asserts. Pairing
// the i-th print with the i-th expected line left 659 of 7,395 Java cases
// without any assertion at all.
//
// It prints its own diagnostic BEFORE throwing, on the real `System.out` rather
// than into the buffer. That is not decoration: an uncaught error renders as
// `RuntimeError: [object]` under Vybe, so the expected and actual values would
// otherwise be lost entirely.
public class Check {
    // A static String, NOT a StringBuilder. Calling a method on a bare static
    // FIELD receiver fails under Vybe with "undefined is not callable"
    // (measured): `SB.append(x)` throws while `StringBuilder l = SB;
    // l.append(x)` works, so the method is resolved from the receiver's
    // declared type at the call site and a static field carries none. String
    // concatenation onto a static field has no such problem.
    static String __buf = "";

    static void __p(Object o) {
        __buf = __buf + String.valueOf(o) + "\n";
    }

    static void __pr(Object o) {
        __buf = __buf + String.valueOf(o);
    }

    static void __check(String want) {
        String got = __buf;
        // The final `println` contributes a trailing newline that the expected
        // line vector never carried, so it is not part of the comparison.
        if (got.endsWith("\n")) {
            got = got.substring(0, got.length() - 1);
        }
        if (!got.equals(want)) {
            System.out.println("FAIL: want [" + want + "] got [" + got + "]");
            throw new RuntimeException("assertion failed");
        }
    }
}

File: crates/testrunner/harness/dart/check.dart
// Vybe test harness — Dart.
//
// Real Dart: this file analyses and runs under the Dart SDK on its own, which
// is what lets an extracted test be compared against the reference
// implementation.
//
// Output is COLLECTED, not paired. The emitter rewrites every `print(x)` in
// the test body into `__p(x)` and compares the whole output once, so a program
// whose print count is not static — a loop — still asserts. Pairing the i-th
// print with the i-th expected line cannot express that, and loops are the
// largest unpairable category in every language where it was tried.
//
// The test's own `main` is renamed to `__vybeMain` and called from a wrapper,
// which avoids having to find the closing brace of `main` to append the check.
// The wrapper is always `async` and always `await`s: `await` on a non-Future
// is legal Dart and returns the value unchanged, so one shape covers both
// `void main()` and `Future<void> main() async`.
//
// `print` here is the REAL print, deliberately: only the body's prints are
// rewritten, so the diagnostic still reaches stdout. It is emitted BEFORE the
// throw because an uncaught error renders as `RuntimeError: [object]` under
// Vybe, which would lose the expected and actual values entirely.

final StringBuffer __vybeOut = StringBuffer();

void __p(Object? o) {
  __vybeOut.writeln(o);
}

void __check(String want) {
  var got = __vybeOut.toString();
  // `writeln` on the final print contributes a trailing newline that the
  // expected line vector never carried.
  if (got.endsWith('\n')) {
    got = got.substring(0, got.length - 1);
  }
  if (got != want) {
    print('FAIL: want [$want] got [$got]');
    throw Exception('assertion failed');
  }
}
